overlord/snapstate/target.go

```diff
@@ -0,0 +1,552 @@
+
+	installables, err := target.Installables(ctx, st, snaps, opts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// TODO: should this be a field on the targets?
+	installInfos := make([]minimalInstallInfo, 0, len(installables))
+	for _, target := range installables {
+		installInfos = append(installInfos, installSnapInfo{target.Info})
+	}
+
+	if err = checkDiskSpace(st, "install", installInfos, opts.UserID, opts.PrereqTracker); err != nil {
+		return nil, nil, err
+	}
+
+	tasksets := make([]*state.TaskSet, 0, len(installables))
+	infos := make([]*snap.Info, 0, len(installables))
+	for _, inst := range installables {
+		if inst.Snap.SnapPath != "" && inst.Snap.DownloadInfo != nil {
+			return nil, nil, errors.New("internal error: installable cannot specify both a path and a download info")
+		}
+
+		info := inst.Info
+
+		if opts.Flags.RequireTypeBase && info.Type() != snap.TypeBase && info.Type() != snap.TypeOS {
+			return nil, nil, fmt.Errorf("unexpected snap type %q, instead of 'base'", info.Type())
+		}
+
+		opts.PrereqTracker.Add(info)
+
+		snapst, ok := snaps[info.InstanceName()]
+		if !ok {
+			snapst = &SnapState{}
+		}
+
+		flags, err := earlyChecks(st, snapst, info, opts.Flags)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		providerContentAttrs := defaultProviderContentAttrs(st, info, opts.PrereqTracker)
+		snapsup := &SnapSetup{
+			Channel:      inst.Snap.Channel,
+			DownloadInfo: inst.Snap.DownloadInfo,
+			SnapPath:     inst.Snap.SnapPath,
+			CohortKey:    inst.Snap.CohortKey,
+
+			Base:               info.Base,
+			Prereq:             getKeys(providerContentAttrs),
+			PrereqContentAttrs: providerContentAttrs,
+			UserID:             opts.UserID,
+			Flags:              flags.ForSnapSetup(),
+			SideInfo:           &info.SideInfo,
+			Type:               info.Type(),
+			Version:            info.Version,
+			PlugsOnly:          len(info.Slots) == 0,
+			InstanceKey:        info.InstanceKey,
+			ExpectedProvenance: info.SnapProvenance,
+		}
+
+		var instFlags int
+		if opts.Flags.SkipConfigure {
+			instFlags |= skipConfigure
+		}
+
+		ts, err := doInstall(st, snapst, snapsup, instFlags, opts.FromChange, inUseFor(opts.DeviceCtx))
+		if err != nil {
+			return nil, nil, err
+		}
+
+		ts.JoinLane(laneGenerator(st, opts))
+
+		tasksets = append(tasksets, ts)
+		infos = append(infos, info)
+	}
+
+	return infos, tasksets, nil
+}
+
+func laneGenerator(st *state.State, opts Options) int {
+	// If transactional, use a single lane for all snaps, so when
+	// one fails the changes for all affected snaps will be
+	// undone. Otherwise, have different lanes per snap so failures
+	// only affect the culprit snap.
+	switch opts.Flags.Transaction {
+	case client.TransactionAllSnaps:
+		return opts.Flags.Lane
+	case client.TransactionPerSnap:
+		return st.NewLane()
+	}
+	return 0
+}
+
+func setDefaultSnapstateOptions(st *state.State, opts *Options) error {
+	var err error
+	opts.DeviceCtx, err = DevicePastSeeding(st, opts.DeviceCtx)
```

Comment:
I also think we probably want to express this difference differently now that we have Options, I think the issue is more first boot code vs later code

Reply:
Are you thinking we should add some field to Options that indicates that the system doesn't need to be seeded? Then it could be used in the first boot code?